src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp

```diff
@@ -461,342 +424,120 @@ RECT NonClientIslandWindow::GetMaxWindowRectInPixels(const RECT* const prcSugges
-            rcMaximum = GetMaxWindowRectInPixels(&rcCurrent, &dpiOfMaximum);
-        }
-        else
-        {
-            // In other circumstances, assume we're snapping around or some
-            // other jump (TS). Just do whatever we're told using the new
-            // suggestion as the restriction monitor.
-            rcMaximum = GetMaxWindowRectInPixels(&rcSuggested, &dpiOfMaximum);
-        }
+// Method Description:
+// - Updates the window frame's theme depending on the application theme (light
+//   or dark). If this is called after the frame is rendered, the old frame is
+//   not invalidated so the frame doesn't update until the user resizes or
+//   focuses/unfocuses the window.
+// Return Value:
+// - <none>
+void NonClientIslandWindow::_UpdateFrameTheme() const
+{
+    const auto isDarkMode = Application::Current().RequestedTheme() == ApplicationTheme::Dark;
 
-        const auto suggestedWidth = szSuggested.cx;
-        const auto suggestedHeight = szSuggested.cy;
-
-        const auto maxWidth = RECT_WIDTH(&rcMaximum);
-        const auto maxHeight = RECT_HEIGHT(&rcMaximum);
-
-        // Only apply the maximum size restriction if the current DPI matches
-        // the DPI of the maximum rect. This keeps us from applying the wrong
-        // restriction if the monitor we're moving to has a different DPI but
-        // we've yet to get notified of that DPI change. If we do apply it, then
-        // we'll restrict the console window BEFORE its been resized for the DPI
-        // change, so we're likely to shrink the window too much or worse yet,
-        // keep it from moving entirely. We'll get a WM_DPICHANGED, resize the
-        // window, and then process the restriction in a few window messages.
-        if (((int)dpiOfMaximum == _currentDpi) &&
-            ((suggestedWidth > maxWidth) ||
-             (suggestedHeight > maxHeight)))
-        {
-            RECT frame{};
-            // Calculate the maxmized window overhang by getting the size of the window frame.
-            // We use the style without WS_CAPTION otherwise the caption height is included.
-            // Only remove WS_DLGFRAME since WS_CAPTION = WS_DLGFRAME | WS_BORDER,
-            // but WS_BORDER is needed as it modifies the calculation of the width of the frame.
-            const auto targetStyle = windowStyle & ~WS_DLGFRAME;
-            AdjustWindowRectExForDpi(&frame, targetStyle, false, GetWindowExStyle(_window.get()), _currentDpi);
-
-            // Frame left and top will be negative
-            _maximizedMargins.cxLeftWidth = frame.left * -1;
-            _maximizedMargins.cyTopHeight = frame.top * -1;
-            _maximizedMargins.cxRightWidth = frame.right;
-            _maximizedMargins.cyBottomHeight = frame.bottom;
-
-            _isMaximized = true;
-            THROW_IF_FAILED(_UpdateFrameMargins());
-        }
-    }
-    else
+    if (FAILED(ThemeUtils::SetDwmImmersiveDarkMode(_window.get(), isDarkMode)))

```

Comment:
this one should just be `LOG_IF_FAILED(...)` -- not because it's fewer lines of code, but because this function returns an HRESULT, and DWM returns an HRESULT, but `LOG_LAST_ERROR` uses `GetLastError()` to get the _win32 error state_ (which isn't used by any of the downstream callees here)

LOG_IF_FAILED() just takes the HRESULT return and logs that. :smile:

Reply:
It actually will shout at you if you use it wrong, but that would require us to hit a failure:

```
                WI_USAGE_ERROR_FORWARD("CALLER BUG: Macro usage error detected.  GetLastError() does not have an error.");
```